Firestore/core/src/firebase/firestore/remote/stream.h

```diff
@@ -0,0 +1,247 @@
+
+namespace firebase {
+namespace firestore {
+namespace remote {
+
+/**
+ * A `Stream` is an abstract base class that represents a bidirectional
+ * streaming connection to the Firestore backend. It's built on top of gRPC C++
+ * library and adds several critical features for our clients:
+ *
+ *   - Exponential backoff on failure (independent of the gRPC mechanism)
+ *   - Authentication via CredentialsProvider
+ *   - Dispatching all callbacks into the shared Firestore async queue
+ *   - Closing idle streams after 60 seconds of inactivity
+ *
+ * Subclasses of `Stream`:
+ *
+ *   - Implement serialization and deserialization of protocol buffers
+ *   - Notify their delegate about stream open/read/error events.
+ *
+ * ## Starting and Stopping
+ *
+ * Streams are stateful and need to be `Start`ed before messages can
+ * be sent and received. A `Stream` can be started and stopped repeatedly.
+ */
+class Stream : public GrpcStreamObserver,
+               public std::enable_shared_from_this<Stream> {
+ public:
+  Stream(util::AsyncQueue* async_queue,
+         auth::CredentialsProvider* credentials_provider,
+         Datastore* datastore,
+         util::TimerId backoff_timer_id,
+         util::TimerId idle_timer_id);
+
+  /**
+   * Starts the stream. Only allowed if `IsStarted` returns false. The stream is
+   * not immediately ready for use: `OnStreamStart` will be invoked when the
+   * stream is ready for outbound requests, at which point `IsOpen` will return
+   * true.
+   *
+   * When start returns, `IsStarted` will return true.
+   */
+  void Start();
+
+/**
+   * Stops the stream. This call is idempotent and allowed regardless of the
+   * current `IsStarted` state.
+   *
+   * When stop returns, `IsStarted` and `IsOpen` will both return false.
+   */
+  void Stop();
+
+  /**
+   * Returns true if `Start` has been called and no error has occurred. True
+   * indicates the stream is open or in the process of opening (which
+   * encompasses respecting backoff, getting auth tokens, and starting the
+   * actual stream). Use `IsOpen` to determine if the stream is open and ready
+   * for outbound requests.
+   */
+  bool IsStarted() const;
+
+  /**
+   * Returns true if the underlying stream is open (`OnStreamStart` has been
+   * called) and the stream is ready for outbound requests.
+   */
+  bool IsOpen() const;
+
+  // `GrpcStreamObserver` interface -- do not use.
+  void OnStreamStart() override;
+  void OnStreamRead(const grpc::ByteBuffer& message) override;
+  void OnStreamError(const util::Status& status) override;
+
+  /**
+   * After an error, the stream will usually back off on the next attempt to
+   * start it. If the error warrants an immediate restart of the stream, the
+   * sender can use this to indicate that the receiver should not back off.
+   *
+   * Each error will call `OnStreamClose`. That function can decide to
+   * cancel backoff if required.
+   */
+  void CancelBackoff();
```

Comment:
Forgot to mention: other platforms use `InhibitBackoff`. I'm not sure why "inhibit" was chosen -- to me, it seems to imply something less than cancellation, as if backoff time was reduced or similar.

Reply:
There aren't any great words for this. We generally use "cancel" to mean that we're stopping some thing that's in flight. That's not the case here and I don't think this is right.

We used "inhibit" here because this is suppressing the next backoff.

ResetBackoff?

Inhibit isn't a bad name on the whole and has the benefit that it doesn't require us to port the name change to the other platform.

Reply:
After offline discussion, reverted to `InhibitBackoff`.